openpype/plugins/publish/integrate_new.py

```diff
@@ -110,801 +156,494 @@ class IntegrateAssetNew(pyblish.api.InstancePlugin):
-                # could throw exception, will be caught in 'process'
-                # all integration to DB is being done together lower,
-                # so no rollback needed
-                self.log.debug("Integrating source files to destination ...")
-                self.integrated_file_sizes.update(self.integrate(instance))
-                self.log.debug("Integrated files {}".
-                               format(self.integrated_file_sizes))
-
-            # get 'files' info for representation and all attached resources
-            self.log.debug("Preparing files information ...")
-            representation["files"] = self.get_files_info(
-                instance,
-                self.integrated_file_sizes)
-
-            self.log.debug("__ representation: {}".format(representation))
-            destination_list.append(dst)
-            self.log.debug("__ destination_list: {}".format(destination_list))
-            instance.data['destination_list'] = destination_list
-            representations.append(representation)
-            published_representations[repre_id] = {
-                "representation": representation,
-                "anatomy_data": template_data,
-                "published_files": published_files
-            }
-            self.log.debug("__ representations: {}".format(representations))
-
-        # Remove old representations if there are any (before insertion of new)
-        if existing_repres:
-            repre_ids_to_remove = []
-            for repre in existing_repres:
-                repre_ids_to_remove.append(repre["_id"])
-            io.delete_many({"_id": {"$in": repre_ids_to_remove}})
-
-        for rep in instance.data["representations"]:
-            self.log.debug("__ rep: {}".format(rep))
-
-        io.insert_many(representations)
-        instance.data["published_representations"] = (
-            published_representations
+            # todo: reduce/simplify what is returned from this function
+            prepared = self.prepare_representation(repre,
+                                                   template_name,
+                                                   existing_repres_by_name,
+                                                   version,
+                                                   instance_stagingdir,
+                                                   instance)
+
+            for src, dst in prepared["transfers"]:
+                # todo: add support for hardlink transfers
+                file_transactions.add(src, dst)
+
+            prepared_representations.append(prepared)
+
+        if not prepared_representations:
+            # Even though we check `instance.data["representations"]` earlier
+            # this could still happen if all representations were tagged with
+            # "delete" and thus are skipped for integration
+            raise RuntimeError("No representations prepared to publish.")
```

Comment:
But that may be intended and is not reason to raise exception.

Reply:
Thanks, good point!

I don't think we would want versions integrated in the database without **any** representations or files. Or what use case would having no representations to integrate have whatsoever? Would be good to have a use case so I could add it to the comment when that might be needed and could be allowed - that way it remains clear in the code why something was designed a certain way.

Or do you mean that instead or erroring it would suffice to "skip the integrator" completely for that instance and just log a warning that says something like "No representations to publish. Integration skipped..."?

Reply:
Looking at what @mkolar writes [here](https://github.com/pypeclub/OpenPype/pull/2898#discussion_r840424267) deleted representations should be for versions that would still be writing e.g. a Prores file and integrating that instead. Which means that if NO representations are detected it's still the unexpected result and thus an error is valid. If e.g. the prores wasn't marked for delete then this would pass fine and things work as expected.

Reply:
I'd say that instead of error, we should be skipping integration in case of no representations present. From developer and TD standpoint you want to have the exception to make sure nothing is missed magically, however, this is mostly encounter by users (that's if the situation happens at all of course) in which case any error that they cannot affect and doesn't really break anything just invokes a small panic every time.

Reply:
I'll log a warning and return early. We'll need to make sure other plug-ins afterwards don't do odd stuff then of course.
Or would you rather have it log just an info message?

Reply:
Just had to handle that situation and moved filtering of representations as first step before anything happens in `IntegrateAssetNew` in [PR](https://github.com/pypeclub/OpenPype/pull/3398).

Reply:
Great. That should resolve it - both old vs. new integrator now solve it by raising an error if I'm not mistaken.